File: Frontend/src/types/contracts/coding.ts
import { z } from "zod";

export const codingDifficultySchema = z.enum(["easy", "medium", "hard"]);
export type CodingDifficulty = z.infer<typeof codingDifficultySchema>;

export const codingLanguageSchema = z.enum(["python", "javascript"]);
export type CodingLanguage = z.infer<typeof codingLanguageSchema>;

export const submissionStatusSchema = z.enum([
  "judging",
  "accepted",
  "wrong_answer",
  "time_limit_exceeded",
  "runtime_error",
  "compile_error",
]);
export type SubmissionStatus = z.infer<typeof submissionStatusSchema>;

export const paramTypeSchema = z.enum([
  "int",
  "float",
  "string",
  "boolean",
  "list_int",
  "list_float",
  "list_string",
  "list_boolean",
  "list_list_int",
  "list_list_string",
  "list_node",
  "tree_node",
  "list_list_node_nullable",
]);
export type ParamType = z.infer<typeof paramTypeSchema>;

export const testCaseSchema = z.object({
  inputArgs: z.array(z.any()),
  expected: z.any().optional(),
  isExample: z.boolean().optional(),
});
export type TestCase = z.infer<typeof testCaseSchema>;

export const functionParamSchema = z.object({
  name: z.string(),
  type: paramTypeSchema,
});
export type FunctionParam = z.infer<typeof functionParamSchema>;

export const problemExampleSchema = z.object({
  input: z.string(),
  output: z.string(),
  explanation: z.string().optional(),
});
export type ProblemExample = z.infer<typeof problemExampleSchema>;

export const problemSummarySchema = z.object({
  slug: z.string(),
  number: z.number(),
  title: z.string(),
  difficulty: codingDifficultySchema,
  topics: z.array(z.string()),
  acceptanceRate: z.number(),
  status: z.enum(["solved", "attempted", "todo"]),
});
export type ProblemSummary = z.infer<typeof problemSummarySchema>;

export const problemDetailSchema = z.object({
  id: z.string(),
  slug: z.string(),
  number: z.number(),
  title: z.string(),
  difficulty: codingDifficultySchema,
  topics: z.array(z.string()),
  descriptionMd: z.string(),
  examples: z.array(problemExampleSchema),
  constraintsMd: z.string(),
  functionName: z.string(),
  params: z.array(functionParamSchema),
  returnType: paramTypeSchema,
  returnIndex: z.number().nullable().optional(),
  starterCode: z.record(z.string(), z.string()),
  testCases: z.array(testCaseSchema),
  timeLimitMs: z.number(),
  editorialMd: z.string(),
  userStatus: z.enum(["solved", "attempted", "todo"]),
});
export type ProblemDetail = z.infer<typeof problemDetailSchema>;

export const problemListResponseSchema = z.object({
  items: z.array(problemSummarySchema),
  total: z.number(),
});
export type ProblemListResponse = z.infer<typeof problemListResponseSchema>;

export const topicCountSchema = z.object({
  name: z.string(),
  count: z.number(),
});
export type TopicCount = z.infer<typeof topicCountSchema>;

export const firstFailureDetailSchema = z.object({
  inputArgs: z.array(z.any()),
  expected: z.any(),
  actual: z.any().optional(),
  debugOutput: z.string().optional(),
  error: z.string().optional(),
});
export type FirstFailureDetail = z.infer<typeof firstFailureDetailSchema>;

export const codingSubmissionSchema = z.object({
  id: z.string(),
  userId: z.string(),
  problemSlug: z.string(),
  language: codingLanguageSchema,
  code: z.string(),
  status: submissionStatusSchema,
  passedCount: z.number(),
  totalCount: z.number(),
  runtimeMs: z.number().nullable().optional(),
  compileStderr: z.string().nullable().optional(),
  firstFailure: firstFailureDetailSchema.nullable().optional(),
  createdAt: z.string(),
});
export type CodingSubmission = z.infer<typeof codingSubmissionSchema>;

export const codingSubmissionSummarySchema = z.object({
  id: z.string(),
  problemSlug: z.string(),
  language: codingLanguageSchema,
  status: submissionStatusSchema,
  passedCount: z.number(),
  totalCount: z.number(),
  runtimeMs: z.number().nullable().optional(),
  createdAt: z.string(),
});
export type CodingSubmissionSummary = z.infer<typeof codingSubmissionSummarySchema>;

export const runTestCaseInputSchema = z.object({
  inputArgs: z.array(z.any()),
});
export type RunTestCaseInput = z.infer<typeof runTestCaseInputSchema>;

export const runCodeRequestSchema = z.object({
  language: codingLanguageSchema,
  code: z.string(),
  testCases: z.array(runTestCaseInputSchema).optional(),
});
export type RunCodeRequest = z.infer<typeof runCodeRequestSchema>;

export const runResultItemSchema = z.object({
  index: z.number(),
  inputArgs: z.array(z.any()),
  expected: z.any().optional(),
  actual: z.any().optional(),
  passed: z.boolean(),
  debugOutput: z.string(),
  error: z.string().nullable().optional(),
  runtimeMs: z.number().nullable().optional(),
});
export type RunResultItem = z.infer<typeof runResultItemSchema>;

export const runCodeResponseSchema = z.object({
  results: z.array(runResultItemSchema),
  compileError: z.string().nullable().optional(),
});
export type RunCodeResponse = z.infer<typeof runCodeResponseSchema>;

export const submitCodeRequestSchema = z.object({
  language: codingLanguageSchema,
  code: z.string(),
});
export type SubmitCodeRequest = z.infer<typeof submitCodeRequestSchema>;

export const submitCodeResponseSchema = z.object({
  submissionId: z.string(),
});
export type SubmitCodeResponse = z.infer<typeof submitCodeResponseSchema>;

export const saveDraftRequestSchema = z.object({
  language: codingLanguageSchema,
  code: z.string(),
});
export type SaveDraftRequest = z.infer<typeof saveDraftRequestSchema>;

export const draftResponseSchema = z.object({
  language: codingLanguageSchema,
  code: z.string(),
  updatedAt: z.string().nullable().optional(),
});
export type DraftResponse = z.infer<typeof draftResponseSchema>;

export const topicStatSchema = z.object({
  topic: z.string(),
  solved: z.number(),
  total: z.number(),
});
export type TopicStat = z.infer<typeof topicStatSchema>;

export const recentSubmissionStatSchema = z.object({
  id: z.string(),
  problemSlug: z.string(),
  problemTitle: z.string(),
  difficulty: codingDifficultySchema,
  status: submissionStatusSchema,
  language: codingLanguageSchema,
  createdAt: z.string(),
});
export type RecentSubmissionStat = z.infer<typeof recentSubmissionStatSchema>;

export const codingStatsSchema = z.object({
  totalSolved: z.number(),
  totalProblems: z.number(),
  easySolved: z.number(),
  easyTotal: z.number(),
  mediumSolved: z.number(),
  mediumTotal: z.number(),
  hardSolved: z.number(),
  hardTotal: z.number(),
  acceptanceRate: z.number(),
  topicStats: z.array(topicStatSchema),
  recentSubmissions: z.array(recentSubmissionStatSchema),
});
export type CodingStats = z.infer<typeof codingStatsSchema>;

File: Backend/app/schemas/coding.py
from typing import Any, ClassVar, Literal

from app.core.serialization import CamelModel
from app.core.timeutils import UtcDatetime
from app.schemas.common import (
    CheckerKind,
    CodingDifficulty,
    CodingLanguage,
    ParamType,
    SubmissionStatus,
)


class TestCase(CamelModel):
    __test__ = False
    input_args: list[Any]
    expected: Any
    is_example: bool = False


class FunctionParam(CamelModel):
    name: str
    type: ParamType


class ProblemExample(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"explanation"})

    input: str
    output: str
    explanation: str | None = None


class CodingProblem(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"return_index"})

    id: str
    slug: str
    number: int
    title: str
    difficulty: CodingDifficulty
    topics: list[str]
    description_md: str
    examples: list[ProblemExample]
    constraints_md: str
    function_name: str
    params: list[FunctionParam]
    return_type: ParamType
    return_index: int | None = None
    checker: CheckerKind = CheckerKind.EXACT
    starter_code: dict[str, str]
    test_cases: list[TestCase]
    time_limit_ms: int = 2000
    editorial_md: str


class ProblemSummary(CamelModel):
    slug: str
    number: int
    title: str
    difficulty: CodingDifficulty
    topics: list[str]
    acceptance_rate: float
    status: Literal["solved", "attempted", "todo"]


class ProblemDetail(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"return_index"})

    id: str
    slug: str
    number: int
    title: str
    difficulty: CodingDifficulty
    topics: list[str]
    description_md: str
    examples: list[ProblemExample]
    constraints_md: str
    function_name: str
    params: list[FunctionParam]
    return_type: ParamType
    return_index: int | None = None
    starter_code: dict[str, str]
    test_cases: list[TestCase]
    time_limit_ms: int
    editorial_md: str
    user_status: Literal["solved", "attempted", "todo"] = "todo"


class ProblemListResponse(CamelModel):
    items: list[ProblemSummary]
    total: int


class TopicCount(CamelModel):
    name: str
    count: int


class FirstFailureDetail(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"actual", "error"})

    input_args: list[Any]
    expected: Any
    actual: Any | None = None
    debug_output: str = ""
    error: str | None = None


class CodingSubmission(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset(
        {"runtime_ms", "compile_stderr", "first_failure"}
    )

    id: str
    user_id: str
    problem_slug: str
    language: CodingLanguage
    code: str
    status: SubmissionStatus
    passed_count: int
    total_count: int
    runtime_ms: int | None = None
    compile_stderr: str | None = None
    first_failure: FirstFailureDetail | None = None
    created_at: UtcDatetime


class CodingSubmissionSummary(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"runtime_ms"})

    id: str
    problem_slug: str
    language: CodingLanguage
    status: SubmissionStatus
    passed_count: int
    total_count: int
    runtime_ms: int | None = None
    created_at: UtcDatetime


class CodeDraft(CamelModel):
    id: str
    user_id: str
    problem_slug: str
    language: CodingLanguage
    code: str
    updated_at: UtcDatetime


class RunTestCaseInput(CamelModel):
    input_args: list[Any]


class RunCodeRequest(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"test_cases"})

    language: CodingLanguage
    code: str
    test_cases: list[RunTestCaseInput] | None = None


class RunResultItem(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset(
        {"expected", "actual", "error", "runtime_ms"}
    )

    index: int
    input_args: list[Any]
    expected: Any | None = None
    actual: Any | None = None
    passed: bool
    debug_output: str = ""
    error: str | None = None
    runtime_ms: int | None = None


class RunCodeResponse(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"compile_error"})

    results: list[RunResultItem]
    compile_error: str | None = None


class SubmitCodeRequest(CamelModel):
    language: CodingLanguage
    code: str


class SubmitCodeResponse(CamelModel):
    submission_id: str


class SaveDraftRequest(CamelModel):
    language: CodingLanguage
    code: str


class DraftResponse(CamelModel):
    omit_if_none: ClassVar[frozenset[str]] = frozenset({"updated_at"})

    language: CodingLanguage
    code: str
    updated_at: UtcDatetime | None = None


class TopicStat(CamelModel):
    topic: str
    solved: int
    total: int


class RecentSubmissionStat(CamelModel):
    id: str
    problem_slug: str
    problem_title: str
    difficulty: CodingDifficulty
    status: SubmissionStatus
    language: CodingLanguage
    created_at: UtcDatetime


class CodingStats(CamelModel):
    total_solved: int
    total_problems: int
    easy_solved: int
    easy_total: int
    medium_solved: int
    medium_total: int
    hard_solved: int
    hard_total: int
    acceptance_rate: float
    topic_stats: list[TopicStat]
    recent_submissions: list[RecentSubmissionStat]
